test/python/primitives/test_backend_sampler.py

```diff
@@ -0,0 +1,565 @@
+from qiskit.exceptions import QiskitError
+from qiskit.primitives import BackendSampler, SamplerResult
+from qiskit.providers import JobStatus, JobV1
+from qiskit.test import QiskitTestCase
+from qiskit.providers.fake_provider import FakeNairobi, FakeNairobiV2
+
+BACKENDS = [FakeNairobi(), FakeNairobiV2()]
+
+
+@ddt
+class TestBackendSampler(QiskitTestCase):
+    """Test BackendSampler"""
+
+    def setUp(self):
+        super().setUp()
+        hadamard = QuantumCircuit(1, 1)
+        hadamard.h(0)
+        hadamard.measure(0, 0)
+        bell = QuantumCircuit(2)
+        bell.h(0)
+        bell.cx(0, 1)
+        bell.measure_all()
+        self._circuit = [hadamard, bell]
+        self._target = [
+            {0: 0.5, 1: 0.5},
+            {0: 0.5, 3: 0.5, 1: 0, 2: 0},
+        ]
+        self._pqc = RealAmplitudes(num_qubits=2, reps=2)
+        self._pqc.measure_all()
+        self._pqc2 = RealAmplitudes(num_qubits=2, reps=3)
+        self._pqc2.measure_all()
+        self._pqc_params = [[0.0] * 6, [1.0] * 6]
+        self._pqc_target = [{0: 1}, {0: 0.0148, 1: 0.3449, 2: 0.0531, 3: 0.5872}]
+        self._theta = [
+            [0, 1, 1, 2, 3, 5],
+            [1, 2, 3, 4, 5, 6],
+            [0, 1, 2, 3, 4, 5, 6, 7],
+        ]
+
+    def _generate_circuits_target(self, indices):
+        if isinstance(indices, list):
+            circuits = [self._circuit[j] for j in indices]
+            target = [self._target[j] for j in indices]
+        else:
+            raise ValueError(f"invalid index {indices}")
+        return circuits, target
+
+    def _generate_params_target(self, indices):
+        if isinstance(indices, int):
+            params = self._pqc_params[indices]
+            target = self._pqc_target[indices]
+        elif isinstance(indices, list):
+            params = [self._pqc_params[j] for j in indices]
+            target = [self._pqc_target[j] for j in indices]
+        else:
+            raise ValueError(f"invalid index {indices}")
+        return params, target
+
+    def _compare_probs(self, prob, target):
+        if not isinstance(prob, list):
+            prob = [prob]
+        if not isinstance(target, list):
+            target = [target]
+        self.assertEqual(len(prob), len(target))
+        for p, targ in zip(prob, target):
+            for key, t_val in targ.items():
+                if key in p:
+                    self.assertAlmostEqual(p[key], t_val, delta=0.1)
+                else:
+                    self.assertAlmostEqual(t_val, 0, delta=0.1)
+
+    @combine(indices=[[0], [1], [0, 1]], backend=BACKENDS)
+    def test_sampler(self, indices, backend):
+        """test for sampler"""
+        circuits, target = self._generate_circuits_target(indices)
+        with self.assertWarns(DeprecationWarning):
+            sampler = BackendSampler(backend=backend, circuits=circuits)
+            result = sampler(list(range(len(indices))), parameter_values=[[] for _ in indices])
```

Comment:
Which of these two lines warns, and how come?  The interfaces have only just been added.

Reply:
TBH, I'm not really sure, I just copied the tests from `test_sampler.py` and updated them to use `BackendSampler` (and also updated them to account for noisy simulation). I'm assuming it's the `__call__` usage, but I'm honestly personally pretty confused by a lot of the interface on primitives it seems weirdly opaque with lots of unexpected calling conventions for me (like putting things like circuits and parameters in the constructor) and a bunch of deprecations for something that was only just recently introduced.

Reply:
This is due to a major change in the specification. Basically, the specification has changed with the development of runtime. Here is a brief history. 
In the first version, the circuit was put in the constructor at session startup, and the `__call__` sent the index of the circuit to reduce the amount of communication. Later, a mechanism was introduced to minimize communication even if circuits were added after constructor. As a result, passing circuits in the constructor and the `__call__` method is deprecated and the run method is used instead.

TBH, personally, I don't want to change the specs so frequently :cry: